crates/workspace/src/workspace.rs

```diff
@@ -3064,6 +3067,23 @@ impl Workspace {
         }))
     }
 
+    pub fn join_pane_into_next(&mut self, pane: View<Pane>, cx: &mut ViewContext<Self>) {
+        let next_pane = self
+            .find_pane_in_direction(SplitDirection::Right, cx)
+            .or_else(|| self.find_pane_in_direction(SplitDirection::Down, cx))
+            .or_else(|| self.find_pane_in_direction(SplitDirection::Left, cx))
+            .or_else(|| self.find_pane_in_direction(SplitDirection::Up, cx));
```

Comment:
I'm conflicted about this. There is already a notion of "next pane" in `activate_next_pane`, meaning "next in the pane group's internal vector, wrapping around at the end". This typically follows the order in which the pane were created.

I tried reusing that notion of "next" for this action, but I found it better UX to have a predictable direction for the join purely based on relative position rather than history. It's also what VSCode does.

Unfortunately this complicates the "close pane if no items left on it". Normally, after closing the current pane, it focuses on the previous pane. But for this action, we need to focus on the pane we're joining into. So we optionally pass the pane to focus on through the close pane flow, which involves several methods and events.

As for joining direction, this order of preference reproduces VSCode's. I'm not particularly attached to it, as long as it's predictable.

Reply:
To me, clockwise order is what I've also imagined.
After playing with it, seems that it's nice to have it this way too, thank you for the explanations.